### PR TITLE
Add a banwords feature

### PR DESCRIPTION
**Pull request by bumbadadabum**

### CODE REVIEW

**Comment by Joimer** [comment]
Great! This is a feature that all bots have because it is necessary.

File: commands.js [hunk 1]
```diff
@@ -1766,6 +1766,77 @@ exports.commands = {
...
+			this.updateBanwords();
+			this.sendReply("Banned phrases succesfully deleted. The list is currently: " + room.banwords.join(', '));
+
+			if (room.chatRoomData) {
+				room.chatRoomData.banwords = room.banwords.slice();
+				Rooms.global.writeChatRoomData();
+			}
+		},
+
+		"": function (target, room, user) {
```
Comment (Joimer) [inline commands.js]: ""?

Comment (bumbadadabum) [inline commands.js]: Just doing /banword. I could change it to /banword view instead, if you'd prefer that.

Comment (Joimer) [inline commands.js]: That's up to preference, what do you think, @Zarel ?

**Comment by SolarisFox** [comment]
It would be nice if there was a little bit of protection against accidentally adding regex banwords that match all messages (like `.*`). I'm not sure what the best way to test this is however.

Also is it possible with this set of code to prevent users from using commands that are sent to the server like `/data` and `/dexsearch`?

**Comment by Slayer95** [comment]
Is this ReDOS safe? @kotarou3 @xfix

**Comment by bumbadadabum** [comment]
@Slayer95 Apparently not. I'll edit this to allow regex only for leaders+.

**Comment by xfix** [comment]
It's not, please don't merge this code. Malicious regex can cause an infinite loop with JS regex engine.

**Comment by TheImmortal** [comment]
I'm personally not a fan of the banwords feature. I don't trust ROs judgement on the banning of words, in addition to not liking the feature itself. If there isn't a discussion thread on this in the Staff forum, make one.

**Comment by bumbadadabum** [comment]
@TheImmortal I do like having the option for people to do this, especially since most bots already have banned phrases. I made the thread btw.

**Comment by Slayer95** [comment]
Also, agreed with TI. Rooms tend to ban words way too "liberally", sometimes with the "it's abused/said too much" reason, which is absolutely invalid imo.

**Comment by Slayer95** [comment]
@bumbadadabum, I wouldn't trust all leaders with regexp banwords. That's only a judgment call appropriately trained technical people can do.

File: commands.js (hunk 1 above)
Comment (Zarel) [inline commands.js]: No comment

**Comment by Zarel** [comment]
I would. Heck, there are some mods I'd trust with regexes. I think at the Leader level it should be expected that you don't add a regex unless you know what you're doing.

**Comment by Zarel** [comment]
@TheImmortal, My issue with opposing a banwords feature is that bots already have it.

If we're going to continue to allow bots to have it, we shouldn't disallow it to be built into the server.

**Comment by bumbadadabum** [comment]
Echoing what @Zarel says.

**Comment by TheImmortal** [comment]
I don't agree with bots having it either, for the record.

Also, that same logic can be applied to a ton of bot features. Most importantly, room blacklists. That should have been a priority over this.

**Comment by bumbadadabum** [comment]
@TheImmortal I don't see how naming blacklists is relevant. Yes, I could have done that but felt like doing this more yesterday. I'll definitely get working on room blacklists at some point, still.

**Comment by bumbadadabum** [comment]
Ok I updated this PR.

**Comment by Slayer95** [comment]
The current implementation still allows <i>any</i> user to DOS the server :|
(Not that I agree now with leaders being able to).

**Comment by bumbadadabum** [comment]
@Slayer95 How so?

**Comment by Slayer95** [comment]
Nvm, didn't check the code awake enough.

File: commands.js [hunk 2]
```diff
@@ -1766,6 +1766,89 @@ exports.commands = {
...
+			this.updateBanwords();
+			this.sendReply("Banned phrases succesfully added. The list is currently: " + room.banwords.join(', '));
+			if (words.length > 1) {
+				this.privateModCommand("The banwords '" + words.join(', ') + "' were added by " + user.name + ".");
+			} else {
+				this.privateModCommand("The banword '" + words[0] + "' was added by " + user.name + ".");
+			}
+
+			if (room.chatRoomData) {
+				room.chatRoomData.banwords = room.banwords.slice();
```
Comment (Zarel) [inline commands.js]: do not `.slice()` this shit

File: commands.js [hunk 3]
```diff
@@ -1766,6 +1766,89 @@ exports.commands = {
...
+			this.updateBanwords();
+			this.sendReply("Banned phrases succesfully deleted. The list is currently: " + room.banwords.join(', '));
+			if (words.length > 1) {
+				this.privateModCommand("The banwords '" + words.join(', ') + "' were removed by " + user.name + ".");
+			} else {
+				this.privateModCommand("The banword '" + words[0] + "' was removed by " + user.name + ".");
+			}
+
+			if (room.chatRoomData) {
+				room.chatRoomData.banwords = room.banwords.slice();
```
Comment (Zarel) [inline commands.js]: ditto

File: commands.js (hunk 1 above)
Comment (Zarel) [inline commands.js]: On second thought, I would prefer `/banword list`. `/banword` alias to `/help banword`

**Comment by bumbadadabum** [comment]
Updated this once again

**Comment by TheImmortal** [comment]
I would have liked it to at least be started as leader only since there was agreement for that from Scotteh and other upper staff.

**Comment by TheImmortal** [comment]
Not your fault bum, that discussion was in the upper staff room.

**Pull request merged**